platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,670 @@
+
+The call to each of these systems is "fire and forget". Specifically, there is no guarantee to Execution that it will
+definitely see an event via `pre-handle` prior to seeing it in `handle`. Technically, Consensus always calls
+`pre-handle` first, but that thread may be parked arbitrarily long by the system and the `handle` thread may actually
+execute first. This is extremely unlikely, but must be defended against in the Execution layer.
+
+### Hashgraph Module
+
+The Hashgraph module orders events into rounds, and assigns timestamps to events. It is given ordered, persisted
+events from Event Intake. Sometimes when an event is added, it turns out to be the last event that was needed to cause
+an entire "round" of events to come to consensus. When this happens, the Hashgraph module emits a `round`. The round
+includes metadata about the round (the list of judge hashes, the round number, the roster, etc.) along with the events
+that were included in the round, in order, with their consensus-assigned timestamps.
+
+![Hashgraph](hashgraph-module.png)
+
+Rounds are immutable. They are sent "fire and forget" style from the Hashgraph module to other modules that require
+them. Some modules only really need the metadata, or a part of the metadata. Others require the actual round data.
+We will pass the full round info (metadata + events) to all listeners, and they can pull from it what they need.
+
+#### Roster Changes
+
+When the roster changes, the Hashgraph algorithm must be made aware. It needs roster information to be able to come
+to consensus. While several modules must respond to roster changes, it is the Hashgraph module that knows the
+applicable roster for any given round. It is given the roster by the Execution layer when it is instructed to produce
+a new round.
+
+At runtime, there is a service in Execution that exposes to the world API for adding nodes to, or removing nodes from,
+the address book. At some point in time, Execution will decide the time is right to construct a new consensus roster,
+and pass that roster to the consensus system. **This must be deterministic**. Each node in the network must assign the
+same roster to the same rounds. The Execution module is able to say exactly which consensus round should use this new
+roster, since it knows about all the rounds already created, and all the rounds it has authorized the Hashgraph module
+to produce, and can therefore know exactly what the next round to produce is, and can tie the new roster to that round.
+
+The Hashgraph module may then apply that roster to that round **or some round in the future**. When a round is produced
+and sent to Execution, the roster used will be supplied. The reason for this is that 3 rounds are involved when coming
+to consensus -- the round that is coming to consensus and 2 newer rounds. Each of these must have been using a
+previously known roster. So when a new roster is supplied, it is **applied** 3 rounds in the future of the currently
```

Comment:
I'm confused what this is saying. Above, it seems to imply that a given roster may be applied to any arbitrary future round. But here, you're saying it's applied 3 rounds in the future.

I'm also confused what you mean by "2 newer rounds". It's not possible to know what round an event will be in if it hasn't yet come to consensus, so why do we need to have the roster for 2 newer rounds?

Reply:
From the Execution perspective, it can be arbitrary (i.e. it doesn't know exactly at what round it will be adopted). But from the Consensus perspective, it has this rounds + 3 methodology. That's what I meant. Does the idea make sense and I just need to fix the wording? Or is the concept off?

Reply:
I don't think I'm grasping the core idea that is being proposed here. 

Are you saying that execution says "use this roster for round `X`", but consensus says "ok, I'll use this for round `X+2`"?

Reply:
I would say more like "Hey, I'd like you to create a new round. And here's a roster you should adopt". And then consensus says "OK". And later it says "that roster you gave me? It is now active in 103".

Reply:
I don't understand what we gain from letting consensus decide what round a roster should be used in.

It makes a lot more sense to me for execution to tell consensus "You can work on round X now, here's the roster you should use"

Reply:
The reason for the consensus module to have a say before was because of the double consensus thing. With that no longer being a factor, it seems needlessly complicated for the consensus module to have a say.

Reply:
🤔 

Let's say that execution has 4 rounds at any given time: 1 round "buffered" and ready to go, 1 round being handled, 1 round being hashed, and 1 round being signed. So let's say round 100 is being signed, 101 is being hashed, 102 is being handled, and 103 is buffered. And let's say that execution controls all of these and Consensus has 0 buffered rounds. The Hashgraph module is literally stalled waiting for Execution to tell it to go ahead with the next round.

Of course, the Hashgraph module has 104 and 105 already in the hashgraph, because 105 was used to collect votes on 103 which is already buffered. When Execution says "it is time for another round, and can you please use this roster going forward", there are already a bunch of events in Consensus that have been buffered up -- there's the events in 104 and 105, but there's also some other events that are in the orphan buffer or whatnot, events that were validated using the old roster.

 Execution doesn't know, or care to know, how many future birth-rounds Consensus happens to be buffering. That's strictly need-to-know information in Consensus! So it cannot know what future round number the new roster is intended for. It just supplies the roster, and Consensus decides what the first available round is, and applies it to that round, and Execution finds out later when it sees a consensus round show up that has the roster.
 
 I don't see another solution without leaking implementation details of Consensus into Execution?

Reply:
If consensus is responsible for deciding what round a roster is for, how is that going to be deterministic across nodes?

Reply:
> there's the events in 104 and 105

You're speaking like we already know which rounds these events are going to be in, but we don't. There are events in the hashgraph that haven't come to consensus yet. Someday, they will end up in rounds 104 and 105. But we won't know until they reach consensus.

 >how many future birth-rounds Consensus happens to be buffering

What do you mean by consensus "buffering" future birth rounds?

Reply:
In order to keep the complexity of the consensus engine from growing, it had been my intention to separate the code for buffering future events from the consensus engine inside a component named the "future event buffer". It works similar to the orphan buffer, except that it holds onto events until they are no longer future events.

Reply:
> If consensus is receiving rosters from execution asynchronously, and is responsible for deciding what round a roster is for, how is that going to be deterministic across nodes?

It isn't asynchronous. The roster received is received deterministically and synchronously. The round applied is up to the consensus system to decide, but is decided deterministically.

> > how many future birth-rounds Consensus happens to be buffering
> What do you mean by consensus "buffering future birth rounds"?

> In order to keep the complexity of the consensus engine from growing, it had been my intention to separate the code for buffering future events from the consensus engine inside a component named the "future event buffer". It works similar to the orphan buffer, except that it holds onto events until they are no longer future events.

I didn't mean that we buffer unspecified future rounds -- in fact, as documented here, I intend to drop those far-future events on the floor. What I mean by "buffering future birth rounds", is holding events that were within `N` future rounds.

So for example, if I am currently working on coming to consensus on round 103, and `N` is 10, then I could have events with birth rounds up to 113 anywhere in the system, including "buffered" in the hashgraph. When I say buffered, I just mean, they are added to the hashgraph but aren't be used for anything particularly at the moment, because I (probably) only need the events that were collected in 104 and 105 to come to consensus on 103, and future birth-round events probably have parents that are newer than those that are close to the round I'm currently working on. That's all I meant by buffered.

Reply:
> When I say buffered, I just mean, they are added to the hashgraph but aren't be used for anything particularly at the moment

This is the exact role of the planned "future event buffer". My main point is that it simplified the architecture a lot making this a stand-alone component, as opposed to integrating it with what we currently call `ConsensusImpl`. There are other parts of the system that need to be protected from future events (mainly event creation), and allowing this to be a separate component prevents the need for duplicated logic and data structures.

Reply:
Thanks @cody-littley for the insight. I am not sure if the "future event buffer" is the same as what I'm thinking, or different, but I will keep this in mind.